planner/optimize.go

```diff
@@ -220,6 +221,25 @@ func isPointGetWithoutDoubleRead(ctx sessionctx.Context, p plannercore.Plan) boo
 	return ok && v.IndexInfo == nil
 }
 
+// OptimizeExecStmt to optimize prepare statement protocol "execute" statement
+// this is a short path ONLY does things filling prepare related params
+// for point select like plan which does not need extra things
+func OptimizeExecStmt(ctx context.Context, sctx sessionctx.Context,
```

Comment:
I think we should check the privilege of the current user, just like what `Optimize()` function did. Can we call it directly?

Reply:
there's a existing problem for cached execution(prepared plan cache and struct plan cached), maybe fix this together in another pull request .

example(prepared plan cache config **enabled**)

| session 1(root) | session 2|
| --- | --- |
| GRANT SELECT ON test.tp TO 'u_tp'@'%' WITH GRANT OPTION;flush privileges; | |
| | prepare ps_stp_r from 'select * from tp where c1 > ?'; set @p1 = 1;|
| |  execute ps_stp_r using @p1; ``` 6 rows in set (0.00 sec)```|
| revoke all on test.tp from 'u_tp';flush privileges;| |
| | execute ps_stp_r using @p1; ``` 6 rows in set (0.00 sec) here not report error```|
| | select * from tp where c1 > 1;  ``` ERROR 1142 (42000): SELECT command denied to user 'u_tp'@'192.168.30.141' for table 'tp``` |

Reply:
plan builder  does not set visitInfo for `execute` statement,   priv error reported in `s.sessionVars.User  != nil`  condition during `buildDataSource`,  a bit difficult to understand..

Reply:
issue https://github.com/pingcap/qa/issues/66 has some explanations on this